src/dune_rules/cram/cram_exec.ml

```diff
@@ -290,16 +290,18 @@ let create_sh_script cram_stanzas ~temp_dir : sh_script Fiber.t =
         ; script = user_shell_code_file
         }
   in
-  fprln oc "trap 'exit 0' EXIT";
+  (* When the main shell script is exited we make sure to clean up any
+     subprocess PIDs as not to orphan them. The first "trap" will kill all
+     processes in the process group and the second "trap" will make sure the main
+     shell process exits gracefully. *)
+  fprln oc {|trap "trap 'exit 0' TERM && kill -- -$$" EXIT|};
```

Comment:
I'd like a little more detail in the comment here for future readers, specifically to understand the reason for registering a trap handler _inside_ another trap handler. Here's my guess at why it's written this way:

The `kill -- -$$` sends `TERM` to all processes in the process group. Is the reason for the "inner" trap so that when said `TERM` is received by the main shell process that it exits with 0? I guess there's no easy way to send `TERM` to all processes in the current process's group _except_ for the current process? And if you were to move the inner trap outside of the `EXIT` trap handler (and just do `kill -- -$$` on `EXIT`) then the process could exit `0` in response to an unexpected `TERM` which we don't want.

Reply:
Exactly, this was my solution to avoiding killing the main process. If we end up killing the main process, then Dune will complain that the process is was running received a kill signal, what we want instead is to preserve the old `trap` behaviour which was to `exit 0`.

By switching the trapping behaviour just before we kill the process group, we can signal to all the processes in the group that they should exit when encountering `TERM`, which they all promptly do.